### DISCUSSION TITLE
Exposing STM board support package & FreeRTOS as a library/module to Fprime?

### DISCUSSION THREAD

**Discussion by BillYJT** [nasa/fprime#1795] [Q&A]
Hi!

I am a member of the Melbourne Space Program. We are trying to use fprime in FreeRTOS on top of a STM32 board. We have been referencing to this [repository](https://github.com/fprime-community/fprime-freertos/tree/v1.0), and we are able to get a build at the moment.

However, the way we include the STM BSP and FreeRTOS is messy -  we added include_directories() in CMakeLists for all FreeRTOS header files in all components calling FreeRTOS/BSP methods. That is, at the moment, the same segment of CMake include_directories() code is in Os/, Svc/, and Ref/. 

In the reference repository, it appears the BSP could be exposed to fprime as a module. However, we haven't been able to do this. So we hope we could receive some guidelines on this.
(P.S. one difference from the reference repo is that we place the STM BSP folder under Os/ rather than under a component in Ref/)

Any input is greatly appreciated!
Best Regards,
Bill.

**Comment by LeStarch** [comment]
We'll gladly help.  Where are y'all getting stuck?

**Reply by LeStarch** [reply]
Here is the generic process for making an F´  package.  

1. Create a repository or folder.
2. Add the path to that repository/folder into your deployment's `settings.ini`

Once this is done, the following will happen:
1. The `cmake` subfolder will be added to CMake such that files can be included as relative "modules"
2. The `cmake/toolchain` and `cmake/platform` subfolders will be exposed to CMake and available to `fprime-util` when choosing the platform/toolchain to build with.
3. The `<folder>/<folder>.cmake` file will automatically be included so as to import custom component

Typically we'd put the `include` statements common to a toolchain in the toolchain file, or in the platform file.  Then they'd automatically be applied to all build steps in the F´  build.

Does this help?  Can I elaborate or expand?

**Reply by BillYJT** [reply]
Thanks for your help @LeStarch !

I tried to set it up following your reply, but it doesn't work at the moment.
I will try to describe my setup below and maybe you can help me pinpoint what I am missing.

**The setup:**
1. I have the entire STM BSP and FreeRTOS under Os/stm32-bsp/
2. I include this folder by having `include_directories(SYSTEM "${FPRIME_FRAMEWORK_PATH}/Os/stm32-bsp/")` in the platform file.
3. In Os/stm32-bsp/ I create a CMakeLists.txt, in which I use `include_directories(...)` to add all the header files and `set(SOURCE_FILES ...)` to add all c files. Also I register the files as a module via `register_fprime_module()`.
4. In Os/CmakeLists.txt, I add `set(MOD_DEPS Os/stm32-bsp)` conditioned to the build target being FreeRTOS, because this is where I need to invoke FreeRTOS method for the OSAL implementation.
5. In the OSAL cpp files, e.g. InterruptLock.cpp, I include the FreeRTOS method by `#include <Os/stm32-bsp/Middlewares/Third_Party/FreeRTOS/Source/CMSIS_RTOS_V2/cmsis_os.h>`

**The error:**
```
/workspaces/fprime/Os/stm32-bsp/Middlewares/Third_Party/FreeRTOS/Source/CMSIS_RTOS_V2/cmsis_os.h:126:10: fatal error: FreeRTOS.h: No such file or directory
  126 | #include "FreeRTOS.h"
```

It seems to me the Os/stm32-bsp/CMakeLists.txt may have gone unnoticed by cmake? As a result of this, the bsp files can't find the relative include path for its own headers?

Let me know if I am not being clear, I am happy to explain more. And thanks again for your guidance.

**Reply by LeStarch** [reply]
You also need to add `Os/stm32-bsp/` to cmake. Otherwise CMake will not run step 3 above.  This can be done with the `CMake` commands: `add_subdirectory` and `include`.  If you are building something you could use `add_subdirectory` and if you are just running commands, use `include`.

I am assuming step 3 has a line of the form `include_directories(path/to/directory/containing/FreeRTOS header)`

**Reply by BillYJT** [reply]
Ok. In my step 2, I am changing `include_directories(SYSTEM "${FPRIME_FRAMEWORK_PATH}/Os/stm32-bsp/")` to `add_subdirectory("${FPRIME_FRAMEWORK_PATH}/Os/stm32-bsp/" "${CMAKE_BINARY_DIR}/F-Prime/Os/stm32-bsp/")` in the platform file. Or do I need both? And is this the correct form for the binary directory?

Now I am getting 
```
CMake Error at /workspaces/fprime/Os/stm32-bsp/CMakeLists.txt:61 (register_fprime_module):
-- Including /workspaces/fprime/cmake/platform/FreeRTOS-stm.cmake
-- Configuring incomplete, errors occurred!
  Unknown CMake command "register_fprime_module".
```

At least it seems the stm32-bsp/CmakeLists.txt is being recognized now. But maybe I am calling register_fprime_module() too early?

**Reply by LeStarch** [reply]
You need both as you have both headers and cmake files.

1. `include_directories` adds a directory for the compiler to search for headers
2. `add_subdirectory` adds a cmake directory to search for cmake modules

In your project there are two cmake include commands: `FPrime.cmake` and `FPrime-Code.cmake`.  Custom setup should go between them and new modules after them.

```
include(..../FPrime.cmake)

# PUT headers here

include(.../Fprime-Code.cmake)

# Put modules here
```

**Reply by BillYJT** [reply]
Thanks again.

Now I remove any include for stm32-bsp/ from the platform file and add in Ref/CMakeLists.txt:
```
include("${CMAKE_CURRENT_LIST_DIR}/../cmake/FPrime.cmake")
# NOTE: register custom targets between these two lines
include_directories("${FPRIME_FRAMEWORK_PATH}/Os/stm32-bsp/")

include("${CMAKE_CURRENT_LIST_DIR}/../cmake/FPrime-Code.cmake")
add_subdirectory("${FPRIME_FRAMEWORK_PATH}/Os/stm32-bsp/" "${CMAKE_BINARY_DIR}/F-Prime/Os/stm32-bsp/")
```

But now getting the following error with `fprime-util generate`:
```
CMake Error at /workspaces/fprime/cmake/module.cmake:31 (add_library):
  add_library cannot create target "Os_stm32-bsp" because another target with
  the same name already exists.  The existing target is a static library
  created in source directory "/workspaces/fprime/Os/stm32-bsp".  See
  documentation for policy CMP0002 for more details.
Call Stack (most recent call first):
  /workspaces/fprime/cmake/module.cmake:80 (generate_base_module_properties)
  /workspaces/fprime/cmake/API.cmake:162 (generate_library)
  /workspaces/fprime/Os/stm32-bsp/CMakeLists.txt:61 (register_fprime_module)


CMake Error at /workspaces/fprime/cmake/prescan.cmake:91 (message):
-- Performing CMake source prescan
-- Configuring incomplete, errors occurred!
  Failed to run prescan, cannot continue.
```

Also attached is the /Os/stm32-bsp/CMakeLists.txt. Should I still use the register_fprime_module() command in this CMakelists?
[CMakeLists.txt](https://github.com/nasa/fprime/files/10174593/CMakeLists.txt)

**Reply by BillYJT** [reply]
Sorry, please ignore my last reply, there is another add_fprime_subdirectory to the same stm32-bsp/ folder I didn't get rid of properly.

I also moved stm32-bsp/ to be on the same level as Os/, Svc/, etc, just to be safe.
Now I can get through `fprime-util generate` alright, but when `fprime-util build`, I got the same error again as in my first question:
```
[ 36%] Building CXX object F-Prime/Os/CMakeFiles/Os.dir/FreeRTOS-stm/InterruptLock.cpp.obj
In file included from /workspaces/fprime/Os/FreeRTOS-stm/InterruptLock.cpp:3:
/workspaces/fprime/stm32-bsp/Middlewares/Third_Party/FreeRTOS/Source/CMSIS_RTOS_V2/cmsis_os.h:126:10: fatal error: FreeRTOS.h: No such file or directory
  126 | #include "FreeRTOS.h"
      |          ^~~~~~~~~~~~
```

This is after stm32-bsp/ being built as a module (0-9%), but when it tries to build OSAL, it looks as if it wants to build stm32-bsp/ again.
Btw, the folder structure inside stm32-bsp/ is quite complex, as I took it from the STM IDE. It's not a flat .h & .cc structure. Could this be an issue?
